### ISSUE TITLE
ARM64: bug in arithmetic

### ISSUE THREAD

**Issue by memcorrupt** [LuaJIT/LuaJIT#918]
Happens on ARM64. LuaJIT (564147f518af5a5d8985d9e09fc3a768231f4e75) compiled with `make` (no other flags) on macOS 13.0.1.

The fourth assertion fails.

```lua
--ieee754 components to double
--sign * (2 ^ (exp - 1023)) * (mantissa / _2pow52 + normal)
local _2pow52 = 2 ^ 52
local a = 1 * (2 ^ (1083 - 1023)) * (4080546448249347 / _2pow52 + 1) --2197541395358679800
local b = -1 * (2 ^ (1052 - 1023)) * (3927497732209973 / _2pow52 + 1) ---1005065126.3690554

assert(a == 2197541395358679800)
assert(b == -1005065126.3690554)
assert(2197541395358679800 % -1005065126.3690554 == -606337536)
assert(a % b == -606337536)```

**Comment by MikePall** [comment]
You're seeing the difference between a fused multiply-add (FMA) operation (or multiply-sub in this case) and an unfused one. The third and fourth assert compare against the unfused result you're apparently expecting.

I'm unsure with which compiler and flags you built LuaJIT for ARM64. Probably XCode using Clang for OSX?

For me both the third *and* the fourth assert fail on ARM64 Linux with GCC. If only the fourth fails, then your C compiler is not generating an ARM64 FMA operation in `lj_vm_foldarith()`, which is tested in the third assert.

The ARM64 interpreter and JIT-compiler uses FMA. The x64 port doesn't, because it doesn't use AVX2 FMA (yet).

In fact this is a problem one has to deal with even in plain C:
```C#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main(int argc, char **argv)
{
  double a = strtod(argv[1], NULL);
  double b =  strtod(argv[2], NULL);
  printf("%.30g\n", a - floor(a / b) * b);
  return 0;
}
```

Compile this on x64 with: `gcc -O3 -mavx2 -mno-fma -o test_unfused test.c -lm` and `gcc -O3 -mavx2 -mfma -o test_fused test.c -lm`
Let's check the results:
```
$ ./test_unfused 2197541395358679800 -1005065126.3690554
-606337536
$ ./test_fused 2197541395358679800 -1005065126.3690554
-606337416.925321817398071289062
```
Note that `-mavx2 -mfma` is the default, if one specifies a modern CPU architecture with `-march=X`.

Unsurprisingly, compiling this for an ARM64 target generates the FMA operation and returns the fused result. Compiling this with `-O0` doesn't and returns the unfused result.

The precise result would be -606337446.0072964. Arguably, the fused result is closer to that.

This raises a couple of questions about whether FMA should always/sometimes/never be used and whether FP results should be cross-platform identical or not.

It can certainly screw up consistency between the JIT compiler and the interpreter, because the JIT compiler may discover more opportunities for fusing. This can lead to various issues.

Apparently there are no easy answers. Given the above tests with GCC, I'm not sure there's even a proper answer for C.

I'm open to (practical) suggestions.

**Comment by corsix** [comment]
> I'm open to (practical) suggestions.

FWIW, I'm _strongly_ of the view that `a * b + c` written in source code should _never_ get turned into FMA; if you want FMA, then you need to explicitly write some inline assembly or invoke an intrinsic or [call a function](https://en.cppreference.com/w/c/numeric/math/fma). For GCC, this means specifying `-ffp-contract=off`.

**Comment by MikePall** [comment]
[Here's](https://stackoverflow.com/a/43357638) an explanation of the current situation with C89, C99, C11 and GCC vs. Clang. Let's just say it's messy.

The options for LuaJIT are:
1. Unconditionally **disable FMA**. Trivial.
2. A **build flag** is easy, too. But you know how this works ... whatever default I pick is wrong for *somebody*. Especially for third-party software that gets linked against a distro-built `libluajit.so`, which may have whatever setting *they* pick.
3. A **runtime flag** is a bit trickier. And it's still a global setting, which is inevitably wrong for *some* module.
4. An **intrinsic or function** sounds cool, but I'm moderately sure even those who *might* benefit from the speedup or the higher precision are not going to bother to add this for every potential line of code. And a lot of FMA opportunities only arise in inlined code after compiler optimizations. Also, good luck implementing that precisely for platforms that don't have it natively (e.g. pre-FMA x64).

That said, simply because of the interpreter vs. JIT difference, I may have no other option than turning it off unconditionally.

Actually, I'd like to have some input from users with a lot of FP-math-heavy code. But I guess they're on x64, which doesn't use FMA (yet).

**Comment by memcorrupt** [comment]
None of these asserts fail on Lua 5.1 (5.1.5 compiled as x86 on Windows)

**Comment by MikePall** [comment]
> None of these asserts fail on Lua 5.1 (5.1.5 compiled as x86 on Windows)

Umm, then try to compile Lua 5.1.5 with GCC for ARM64:
```
$ ./lua-5.1.5-arm64 -e 'print(2197541395358679800 % -1005065126.3690554)'
-606337416.92532
```
That fused result comes from the constant-folding inside the parser. With variables, the interpreter runs it and gives the same result. That's because the C code in `luai_nummod()` is used by both. Which means, the C compiler behavior decides upon the outcome.

While compatibility with Lua 5.1 is important, bug-for-bug compatibility was never a goal.

I do think it's important to make a coherent decision for LuaJIT now. And not leave it up to the architecture, C compiler or C compiler flags.

**Comment by ajojoxiii** [comment]
While I'm aware that LuaJIT is not exactly a native compiler but how about the option 3? The default being the non-fused code but fusion can be enabled via something like '--aggressive-math' optimization flag. Just like gcc/clang do. Users would know that they are doing this at their own risk. Sometimes people do really care about performance and about conformance not so much.

**Comment by estama** [comment]
Hello, i'm the CTO of BeamNG, and we work on simulating vehicles.

TLDR: We are heavy numerical users of LuaJIT, and this kind of numerical differences between the interpreter and the JIT are of no concern to us.

In more detail, and of-course IMHO:

- In the example above i view the == with floating point math as highly suspicious. Equality with floating point math has to be used very carefully, in real code i expect that a check within a range would be happening, or a clamping operation before the check.

- Non-integer numerical code should always be treated as "fuzzy". To give the compiler some useful optimization freedom, we use fast-math on the C++ side, so we are accustomed to this kind of numerical inconsistencies. We also do numerical error analysis of calculations whenever things behave weirdly. This kind of analysis needs to happen a lot more on the C++ side which works with floats rather than on the LuaJIT side which works with doubles.

- We are already accustomed and working around differences between LuaJIT's interpreter and JIT. The most obvious one that we hit all the time, is performance characteristic differences (numerical differences are of little to no concern to us). For example even tho LuaJIT's guidelines say that we shouldn't be doing common sub-expression elimination by hand, we do it because we don't know in which side, interpreter or JIT, a certain code will fall into. The same applies for FFI's performance differences, etc. 

So the reality is that there will be differences between the interpreter and the JIT. Concerning the argument of moving the interpreter and the JIT closer together, instead of disabling optimizations on the JIT side, i would be adding the optimizations to the interpreter (if possible), to reduce the optimization delta.
